File: src/med_exam_toolkit/parsers/yikaobang.py
from __future__ import annotations
from med_exam_toolkit.models import Question, SubQuestion
from med_exam_toolkit.parsers import register
from med_exam_toolkit.parsers.base import BaseParser


@register("yikaobang")
class YikaobangParser(BaseParser):
    """解析 com.yikaobang.yixue 的 JSON 格式"""

    def can_handle(self, raw: dict) -> bool:
        return "yikaobang" in raw.get("pkg", "")

    def parse(self, raw: dict) -> Question:
        mode = raw.get("mode", "")
        q = Question(
            name=raw.get("name", ""),
            pkg=raw.get("pkg", ""),
            cls=raw.get("cls", ""),
            unit=raw.get("unit", ""),
            mode=mode,
            raw=raw,
        )

        if "B" in mode and "型题" in mode:
            q.shared_options = raw.get("shared_options", [])
            q.discuss = raw.get("discuss", "")
            for sq in raw.get("sub_questions", []):
                q.sub_questions.append(SubQuestion(
                    text=sq.get("test", ""),
                    options=q.shared_options,
                    answer=sq.get("answer", ""),
                    rate=sq.get("rate", ""),
                    error_prone=sq.get("error_prone", ""),
                    discuss=sq.get("discuss", ""),
                    point=sq.get("point", ""),
                ))

        elif "A3" in mode or "A4" in mode:
            q.stem = raw.get("test", "")
            for sq in raw.get("sub_questions", []):
                q.sub_questions.append(SubQuestion(
                    text=sq.get("sub_test", ""),
                    options=sq.get("option", []),
                    answer=sq.get("answer", ""),
                    rate=sq.get("rate", ""),
                    error_prone=sq.get("error_prone", ""),
                    discuss=sq.get("discuss", ""),
                    point=sq.get("point", ""),
                ))

        else:
            # A1 / A2 —— yikaobang 多了 point 字段
            q.sub_questions.append(SubQuestion(
                text=raw.get("test", ""),
                options=raw.get("option", []),
                answer=raw.get("answer", ""),
                rate=raw.get("rate", ""),
                error_prone=raw.get("error_prone", ""),
                discuss=raw.get("discuss", ""),
                point=raw.get("point", ""),
            ))

        return q

File: internal/parsers/yikaobang.go
package parsers

import (
	"github.com/zyh001/med-exam-kit/internal/models"
)

// YikaobangParser handles JSON from com.yikaobang.yixue.
type YikaobangParser struct{}

func (p *YikaobangParser) CanHandle(raw map[string]any) bool {
	return contains(str(raw["pkg"]), "yikaobang")
}

func (p *YikaobangParser) Parse(raw map[string]any) (*models.Question, error) {
	mode := str(raw["mode"])
	q := &models.Question{
		Name: str(raw["name"]),
		Pkg:  str(raw["pkg"]),
		Cls:  str(raw["cls"]),
		Unit: str(raw["unit"]),
		Mode: mode,
		Raw:  raw,
	}

	hasSub := len(anySlice(raw["sub_questions"])) > 0

	switch {
	case contains(mode, "B") && contains(mode, "型题") && hasSub:
		// Multi-question B-type with shared options
		q.SharedOptions = strSlice(raw["shared_options"])
		q.Discuss = str(raw["discuss"])
		for _, item := range anySlice(raw["sub_questions"]) {
			sq := item.(map[string]any)
			q.SubQuestions = append(q.SubQuestions, models.SubQuestion{
				Text:       str(sq["test"]),
				Options:    strSliceCopy(q.SharedOptions),
				Answer:     str(sq["answer"]),
				Rate:       str(sq["rate"]),
				ErrorProne: str(sq["error_prone"]),
				Discuss:    str(sq["discuss"]),
				Point:      str(sq["point"]),
			})
		}

	case (contains(mode, "A3") || contains(mode, "A4") || contains(mode, "案例")) && hasSub:
		// Standard multi-subquestion case
		q.Stem = str(raw["test"])
		for _, item := range anySlice(raw["sub_questions"]) {
			sq := item.(map[string]any)
			q.SubQuestions = append(q.SubQuestions, models.SubQuestion{
				Text:       str(sq["sub_test"]),
				Options:    strSlice(sq["option"]),
				Answer:     str(sq["answer"]),
				Rate:       str(sq["rate"]),
				ErrorProne: str(sq["error_prone"]),
				Discuss:    str(sq["discuss"]),
				Point:      str(sq["point"]),
			})
		}

	case contains(mode, "A3") || contains(mode, "A4") || contains(mode, "案例"):
		// Single-question fallback for A3/A4 label
		q.Stem = str(raw["test"])
		q.SubQuestions = append(q.SubQuestions, models.SubQuestion{
			Text:       "",
			Options:    strSlice(raw["option"]),
			Answer:     str(raw["answer"]),
			Rate:       str(raw["rate"]),
			ErrorProne: str(raw["error_prone"]),
			Discuss:    str(raw["discuss"]),
			Point:      str(raw["point"]),
		})

	default: // A1 / A2 / single B-type
		q.SubQuestions = append(q.SubQuestions, models.SubQuestion{
			Text:       str(raw["test"]),
			Options:    strSlice(raw["option"]),
			Answer:     str(raw["answer"]),
			Rate:       str(raw["rate"]),
			ErrorProne: str(raw["error_prone"]),
			Discuss:    str(raw["discuss"]),
			Point:      str(raw["point"]),
		})
	}
	return q, nil
}
